DSP: add dsp support

SiyuanCheng-CN opened the pull request:
add dsp config to enable dsp registers

This commit is just a basic dsp support, DSP SHARING is not guaranteed to fully succeed.

ruuddw commented on arch/Kconfig:
```diff
@@ -742,6 +747,16 @@ config FPU_SHARING
 
 endmenu
 
+menu "Floating Point Options"
```
Why is this under "Floating Point Options"?
The DSP stuff we have is about Fixed Point.

SiyuanCheng-CN replied:
fixed

ruuddw commented on arch/arc/core/offsets/offsets.c:
```diff
@@ -115,6 +115,17 @@ GEN_OFFSET_SYM(_callee_saved_stack_t, dpfp2l);
 #endif
+#endif
+#ifdef CONFIG_DSP
+GEN_OFFSET_SYM(_callee_saved_stack_t, dsp_ctrl);
```
Does it make sense to save the control registers (DSP_CTRL and DSP_FFT_CTRL)?
I guess it depends on the application, but the rounding/saturation etc. settings may be a global choice, not per task.

SiyuanCheng-CN replied:
I'm not quite familiar with DSP. But I think If two threads are running two different DSP applications, DSP_CTRL is quite likely different. And Jacco in MLI team told me that, MLI sets the dsp mode on every call, make sure that at least accumulator registers and dsp mode are saved and restored.

ruuddw replied:
ok, if also DSP mode is used dynamically and not statically configured, then it should be saved. Maybe ask @JaccovG to review this PR as well?

ruuddw commented on include/zephyr/kernel.h:
```diff
@@ -205,6 +205,13 @@ extern void k_thread_foreach_unlocked(
+/**
+ * @brief DSP registers are managed by context switch
+ */
+#define K_DSP_REGS (BIT(5))
```
I don't see any code (not even test code) that's setting/clearing this bit. Shouldn't there be some?

SiyuanCheng-CN replied:
Added DSP_SHARING disable & enable API functions.

ruuddw commented on soc/arc/snps_nsim/Kconfig:
```diff
@@ -9,6 +9,7 @@ choice
 config SOC_NSIM_EM
 	bool "Synopsys ARC EM in nSIM"
 	select CPU_HAS_MPU
+	select CPU_HAS_DSP
```
I think we should introduce a new nSIM target for DSP, to ensure we have coverage with and without DSP extensions.

SiyuanCheng-CN replied:
Will add it and test case in next commit

ruuddw commented on soc/arc/snps_nsim/Kconfig:
```diff
@@ -9,6 +9,7 @@ choice
+	select CPU_HAS_DSP
 
 config SOC_NSIM_EM7D_V22
 	bool "Synopsys ARC EM7D_V22 in nSIM"
```
EM7D is a DSP variant, this one should have the HAS_DSP enabled.

SiyuanCheng-CN replied:
Added.

ruuddw reviewed (changes requested):
Looks good in general, see comments.
Also, I'd like to see some test application added that tests the DSP context switching.

evgeniy-paltsev commented on arch/arc/Kconfig:
```diff
@@ -105,6 +105,11 @@ endif #ISA_ARCV3
 config FP_FPU_DA
 	bool
 
+config ARC_DSP_COMPLEX
```
Let's add help here with the information about what type of DSP HW this option implies.
Note that ARC has various DSP extensions and the generic name like `ARC_DSP_COMPLEX` doesn't help here :)

evgeniy-paltsev commented on arch/arc/Kconfig:
```diff
@@ -105,6 +105,11 @@ endif #ISA_ARCV3
+config ARC_DSP_COMPLEX
+	bool
+	depends on DSP
+	default y
```
Hmm, so we can enable it for all ARC targets. Is it really supported for both EM, ARCv2 HS, ARCv3 HS? 
If no - let's limit it to the exact processor type.

SiyuanCheng-CN replied:
fiexed

evgeniy-paltsev commented on arch/arc/include/swap_macros.h:
```diff
@@ -104,7 +125,7 @@
 	ld_s r13, [r2, ___thread_base_t_user_options_OFFSET]
 	/* K_FP_REGS is bit 1 */
-	bbit0 r13, 1, 2f
+	bbit0 r13, 1, 3f
```
Let's use labels here instead of `2f`, `3f`, `4f, here. It's getting really cryptic with such level of nesting.

SiyuanCheng-CN replied:
fixed

evgeniy-paltsev commented on arch/arc/core/thread.c:
```diff
@@ -253,3 +253,37 @@ int arch_float_enable(struct k_thread *thread, unsigned int options)
+
+	/* Ensure a preemptive context switch does not occur */
+
+	key = irq_lock();
```
Is it OK to use irq_lock and not a spinlock here? Are we expect that this code can be used on SMP platform in the manner that we are calling this function for thread != _current?

SiyuanCheng-CN replied:
fixed

evgeniy-paltsev commented on arch/arc/include/swap_macros.h:
```diff
@@ -121,9 +142,30 @@
+#ifdef CONFIG_DSP_SHARING
+	ld_s r13, [r2, ___thread_base_t_user_options_OFFSET]
+	/* K_DSP_REGS is bit 5 */
+	bbit0 r13, 5, 4f
```
This will break after someone touched the bit definitions in `include/zephyr/kernel.h`
Can we somehow use `K_DSP_REGS` directly here instead of hardcoding the value (5)?

SiyuanCheng-CN replied:
bit(5) is from thread option and I added two functions to set and reset it. Not sure how to configure different threads without using thread option. I think using  `K_DSP_REGS`  macro might affect all threads.

ruuddw commented:
Thanks for adding a test, however it looks like the test needs some more work. It looks like copy-paste-edit of the fpu_sharing, but  with lots of fpu_sharing stuff still in (description etc. refer to floating point, it says it computes pi but it doesn't?).
If the test indeed is very similar, maybe consider modifying/generalizing the FPU sharing test to also cover DSP?
If it is really different, then please clean it up further.

SiyuanCheng-CN commented:
> Thanks for adding a test, however it looks like the test needs some more work. It looks like copy-paste-edit of the fpu_sharing, but with lots of fpu_sharing stuff still in (description etc. refer to floating point, it says it computes pi but it doesn't?). If the test indeed is very similar, maybe consider modifying/generalizing the FPU sharing test to also cover DSP? If it is really different, then please clean it up further.

Yes, it's still a draft, I need to add a test, which includes some DSP instructions to calculate something on XY memory to replace PI calculation. I'm trying to find examples from MLI, but it's not that easy to reproduce it without whole MLI library, do you have any ideas about example? The idea of register context switching from FPU_SHARING is good and also works for DSP register, I'll do more modifications and clean it up.

evgeniy-paltsev commented on tests/kernel/dsp_sharing/generic/src/load_store.c:
```diff
@@ -0,0 +1,251 @@
+		 * IMPORTANT: This logic requires that sys_clock_tick_get_32() not
+		 * perform any dsp operations!
+		 */
+		while (volatile_stack_var--) {
```
We can use `k_busy_wait` for that.

SiyuanCheng-CN replied:
fixed

ruuddw commented on arch/arc/Kconfig:
```diff
@@ -327,6 +327,43 @@ config ARC_NORMAL_FIRMWARE
+	help
+	  This option is to enable Zephyr to store and retore DSP_BFLY0
+	  and FFT_CTRL registers during context switch. This option is
+	  only required when butterfly instrutions are used in
```
typo, instru**c**tions.

ruuddw commented on arch/arc/Kconfig:
```diff
@@ -327,6 +327,43 @@ config ARC_NORMAL_FIRMWARE
+	  This option enables preservation of the hardware AGU registers
+	  across context switches to allow multiple threads to perform concurrent
+	  operations on XY memory. Save and restore small size AGU registers is
+	  set as deault, including 4 address pointers regs, 2 address offset regs
```
de**f**ault

ruuddw commented on arch/arc/Kconfig:
```diff
@@ -327,6 +327,43 @@ config ARC_NORMAL_FIRMWARE
 	  resources of the ARC processors, and, therefore, it shall avoid
 	  accessing them.
 
+config ARC_DSP_COMPLEX
```
naming is a bit inconsistent. No _SHARING for this one?
Furthermore, please consider if this and below options are descriptions of the configured hardware (if so, then _HAS_ may be a better name), or user configuration options to choose to safe (_SHARING_) or not safe the related registers in a context switch.

SiyuanCheng-CN replied:
changed to CONFIG_ARC_DSP_BFLY_SHARING, it is not a configured hardware and it can be configured in `prj.conf` file, just like [dsp test](https://github.com/foss-for-synopsys-dwc-arc-processors/zephyr/blob/c9f9fb4dc9fc994fe419ee962573256bcc790681/tests/kernel/dsp_sharing/generic/prj.conf#:~:text=CONFIG_ARC_DSP_BFLY_SHARING%3Dy).

ruuddw commented on arch/arc/core/offsets/offsets.c:
```diff
@@ -115,8 +115,69 @@ GEN_OFFSET_SYM(_callee_saved_stack_t, dpfp2l);
+GEN_OFFSET_SYM(_callee_saved_stack_t, agu_mod1);
+GEN_OFFSET_SYM(_callee_saved_stack_t, agu_mod2);
+GEN_OFFSET_SYM(_callee_saved_stack_t, agu_mod3);
+#if (!defined CONFIG_ARC_AGU_MEDIUM) || (!defined CONFIG_ARC_AGU_LARGE)
```
is this correct? I'd expect #ifdef CONFIG_ARC_AGU_MEDIUM here.

SiyuanCheng-CN replied:
optimized it

ruuddw commented on cmake/compiler/arcmwdt/compiler_flags.cmake:
```diff
@@ -116,7 +116,7 @@ set_compiler_property(PROPERTY warning_error_misra_sane -Werror=vla)
 set_compiler_property(PROPERTY cstd -std=)
 
 if (NOT CONFIG_ARCMWDT_LIBC)
-  set_compiler_property(PROPERTY nostdinc -Hno_default_include -Hnoarcexlib)
```
Why is no_default_include removed? Don't understand how this relates to this PR. This may have implications beyond DSP as well.

SiyuanCheng-CN replied:
If don't remove -Hno_default_include, `fxarc.h` can't be included. I need to use FXAPI library

ruuddw commented on kernel/include/kernel_arch_interface.h:
```diff
@@ -208,6 +208,39 @@ int arch_float_disable(struct k_thread *thread);
+ *
+ * The function is used to enable the preservation of dsp
+ * context information for a particular thread.
+ * This API depends on each architecture implimentation. If the architecture
```
impl**e**mentation

ruuddw commented on kernel/thread.c:
```diff
@@ -903,6 +903,24 @@ int z_impl_k_float_enable(struct k_thread *thread, unsigned int options)
 
+int z_impl_k_dsp_disable(struct k_thread *thread)
+{
+#if defined(CONFIG_DSP) && defined(CONFIG_DSP_SHARING)
```
the function definition in the header file is conditional, but the implementation is always there and has the condition in the function body. Shouldn't the functions simply not be implemented when DSP is not supported?

SiyuanCheng-CN replied:
Its style is followed from FPU, and I also want to add AGU context enable/disable API. But the left space of thread option is rear and AGU is architecture-specific. I have a CONFIG_AGU_SHARING to control whether to save/load context, but can't control threads separately or enable/disable it during execution( didn't create a thread option, like K_DSP_REGS and API functions). Eugeniy suggested me to [use a macro to replace thread option](https://github.com/foss-for-synopsys-dwc-arc-processors/zephyr/pull/5#:~:text=This%20will%20break%20after%20someone%20touched%20the%20bit%20definitions%20in%20include/zephyr/kernel.h%0ACan%20we%20somehow%20use%20K_DSP_REGS%20directly%20here%20instead%20of%20hardcoding%20the%20value%20(5)%3F). But I think directly use `K_DSP_REGS` can't control threads separately?

ruuddw commented on tests/kernel/dsp_sharing/generic/src/complex_calculation.c:
```diff
@@ -0,0 +1,156 @@
+ * @file
+ * @brief complex number multiplication portion of DSP sharing test
+ *
+ * @ingroup kernel_dspharing_tests
```
kernel_dsp**s**haring_tests

ruuddw commented on tests/kernel/dsp_sharing/generic/src/complex_calculation.c:
```diff
@@ -0,0 +1,156 @@
+ *
+ * Testing utilizes a pair of tasks that independently compute complex vector
+ * dot product. The lower priority task is regularly preempted by the higher
+ * priority task, therebytesting whether DSP context information is properly
```
therebytesting ->  thereby testing

ruuddw commented on arch/arc/Kconfig:
```diff
@@ -327,6 +327,44 @@ config ARC_NORMAL_FIRMWARE
+	depends on DSP && CPU_ARCEM
+	default n
+	help
+	  This option is to enable Zephyr to store and retore DSP_BFLY0
```
typo: reStore

SiyuanCheng-CN replied:
fixed

ruuddw commented on arch/arc/core/thread.c:
```diff
@@ -20,6 +20,7 @@
 #include <zephyr/arch/arc/v2/mpu/arc_core_mpu.h>
 #endif
 
+static struct k_spinlock lock;
```
No conditional compilation for DSP/SHARING?

SiyuanCheng-CN replied:
fixed

ruuddw commented on arch/arc/core/thread.c:
```diff
@@ -253,3 +254,36 @@ int arch_float_enable(struct k_thread *thread, unsigned int options)
 #endif /* CONFIG_FPU && CONFIG_FPU_SHARING */
+
+#if defined(CONFIG_DSP) && defined(CONFIG_DSP_SHARING)
+int arch_dsp_disable(struct k_thread *thread)
```
No 'options' for disable? Wouldn't it be possible that DSP features are still in use, but AGU not?

SiyuanCheng-CN replied:
Added, but don't sure if this case exists.

ruuddw commented on arch/arc/include/swap_macros.h:
```diff
@@ -84,9 +85,129 @@
+dsp_store :
+#ifdef CONFIG_DSP_SHARING
+	ld_s r13, [r2, ___thread_base_t_user_options_OFFSET]
+	/* K_DSP_IDX is 5 */
```
This comment is not needed anymore, as the magic number 5 isn't used.
Same comment below for '7'.
And, for consistency, can you do the same for the FP register above? That still uses '1'.

SiyuanCheng-CN replied:
fixed

ruuddw commented on soc/arc/snps_nsim/CMakeLists.txt:
```diff
@@ -18,7 +18,8 @@ else()
   zephyr_compile_options_ifdef(CONFIG_SOC_NSIM_EM -arcv2em -core3 -Xdiv_rem=radix2
 			       -Xmpy_option=mpyd -Xbitscan -Xswap -Xbarrel_shifter
-			       -Xshift_assist -Xdsp2 -Xdsp_complex
+			       -Xshift_assist -Xdsp2 -Xdsp_complex -Xxy -Xagu_large
```
See earlier comment, I don't think we should enable DSP stuff by default for all nSIM EM targets.

SiyuanCheng-CN replied:
Add a nsim_em_dsp target only for DSP feature

ruuddw commented on arch/Kconfig:
```diff
@@ -564,6 +564,11 @@ config CPU_HAS_DCLS
 	  This option is enabled when the processor hardware is configured in
 	  Dual-redundant Core Lock-step (DCLS) topology.
 
+config CPU_HAS_DSP
```
Shouldn't there be a HAS_AGU as well?

SiyuanCheng-CN replied:
added

ruuddw commented on tests/kernel/dsp_sharing/generic/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+project(dsp_sharing)
+
+zephyr_include_directories(
+      ${ARCMWDT_TOOLCHAIN_PATH}/MetaWare/arc/lib/src/fx/include/
```
this should have some condition to only include for ARC MWDT, or the test should be made ARC specific.

SiyuanCheng-CN replied:
fixed

ruuddw commented on tests/kernel/dsp_sharing/generic/src/complex_calculation.c:
```diff
@@ -0,0 +1,156 @@
+ */
+
+#include <ztest.h>
+#include "fxarc.h"
```
ARC specific, add some #ifdef? Also for the later code that uses the ARC specific features.
Alternatively, make this an arch specific test.

SiyuanCheng-CN replied:
fixed

ruuddw commented on tests/kernel/dsp_sharing/generic/src/complex_calculation.c:
```diff
@@ -0,0 +1,156 @@
+		 * kernel to provide a "clean" DSP state to this thread
+		 * once the sleep ends.
+		 */
+		k_sleep(K_MSEC(10));
```
Preempting here tests the accumulator, are XY and AGU registers covered as well by the test?
E.g. did you test without saving the registers and see if the test failed?

SiyuanCheng-CN replied:
Yes, it failed if don't save&restore AGU registers when attribute data to Xdata region.

ruuddw commented on tests/kernel/dsp_sharing/generic/src/dsp_context.h:
```diff
@@ -0,0 +1,74 @@
+#define _DSPCONTEXT_H
+
+/*
+ * Each architecture must define the following structures (which may be empty):
```
This is good to abstract from CPU details. But the test itself still has dependencies as mentioned in an earlier remark.

ruuddw commented on tests/kernel/dsp_sharing/generic/src/dsp_context.h:
```diff
@@ -0,0 +1,74 @@
+
+struct dsp_volatile_register_set {
+#ifdef CONFIG_ARC_DSP_BFLY_SHARING
+	uintptr_t dsp_bfly0;
```
Is only the butterfly register ARC DSP specific? not the other accumulator and AGU registers?

SiyuanCheng-CN replied:
accumulators(r58, r59) are already been saved when `CONFIG_ARC_HAS_ACCL_REGS` is enabled. Will add some AGU registers to test

ruuddw reviewed (changes requested):
Few comments remaining to be handle before submitting upstream.

ruuddw commented:
Thanks for adding the 'dsp' variant for nsim, but let's call the new config nsim_em11d instead of 'dsp' - that aligns with the ARC core names (see also the EMSK and EMSDP boards). Note that when introducing a new board, also the ARC boards documentation for nSIM simulation should be updated.

SiyuanCheng-CN commented:
> Thanks for adding the 'dsp' variant for nsim, but let's call the new config nsim_em11d instead of 'dsp' - that aligns with the ARC core names (see also the EMSK and EMSDP boards). Note that when introducing a new board, also the ARC boards documentation for nSIM simulation should be updated.

Could you please point the link of documentation that I'm going to amend.

ruuddw commented:
> Could you please point the link of documentation that I'm going to amend.

boards\arc\nsim\doc\index.rst

Commit:
DSP: add DSP support for ARC processor

add DSP reg in context switch
add AGU reg in context switch to support XY mem
add thread option and API to dis/enable DSP switch

Signed-off-by: Siyuan Cheng <siyuanc@synopsys.com>

Commit:
arc: add nsim_em11d target

add nsim_em11d target specific for DSP feature

Signed-off-by: Siyuan Cheng <siyuanc@synopsys.com>

Commit:
DSP: add dsp unit test

add dsp context switch test
add complex multiplication test for ARC processor

Signed-off-by: Siyuan Cheng <siyuanc@synopsys.com>

ruuddw commented on boards/arc/nsim/nsim_em11d.yaml:
```diff
@@ -0,0 +1,12 @@
+identifier: nsim_em11d
+name: EM Nsim simulator
```
"EM11D nSIM simulator" as name would be better.

ruuddw commented on boards/arc/nsim/nsim_em11d.yaml:
```diff
@@ -0,0 +1,12 @@
+simulation: nsim
+arch: arc
+toolchain:
+  - zephyr
```
not sure you want to claim zephyr toolchain support, as the relevant features (DSP) require MWDT.
Did you test with zephyr/gnu toolchain? If things work and e.g. only dsplib or mli is compiled with metaware, but zephyr and applications with gnu, that would be great.

SiyuanCheng-CN replied:
Actually FXAPI library and MLI_v1.1 (current is v2.0) can be compiled with mwdt into static library, then linked to zephyr application with gnu. MLI_v2.0 can't achieve this. It's good to add this in the future, but currently let's just remove gnu, only use mwdt first.

ruuddw commented on soc/arc/snps_nsim/Kconfig:
```diff
@@ -15,6 +15,11 @@ config SOC_NSIM_EM7D_V22
 	select CPU_HAS_FPU
 
+config SOC_NSIM_EM11D
+	bool "Synopsys ARC EM_DSP in nSIM"
```
EM11D

ruuddw commented on soc/arc/snps_nsim/Kconfig:
```diff
@@ -15,6 +15,11 @@ config SOC_NSIM_EM7D_V22
+config SOC_NSIM_EM11D
+	bool "Synopsys ARC EM_DSP in nSIM"
+	select CPU_HAS_MPU
+	select CPU_HAS_DSP
```
no HAS_AGU_REGS here? (I see they are later configured in the em11d specific config - why DSP here, and AGU there?).

SiyuanCheng-CN replied:
My first thinking is DSP is a kind of feature just like FPU or MPU. But AGU is more like registers, just like `CONFIG_ARC_HAS_ACCL_REGS`.

ruuddw replied:
That's ok - the choice is a bit arbitrary. AGU is for me more a feature as well instead of 'more registers', more like a DMA engine (autonomously loads/stores data from registers with auto-incrementing addresses).

ruuddw commented on soc/arc/snps_nsim/Kconfig.defconfig.em11d:
```diff
@@ -0,0 +1,65 @@
+
+if SOC_NSIM_EM11D
+
+config CPU_EM4_FPUDA
```
for EM11D (with caches enabled, as I figured from the nsim properties) the base should actually be EM6. EM4 doesn't have/enable caches).

ruuddw commented on tests/kernel/dsp_sharing/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+zephyr_include_directories(
+      ${ARCMWDT_TOOLCHAIN_PATH}/MetaWare/arc/lib/src/fx/include/
+    )
+target_sources(app PRIVATE src/calculation_arc.c)
```
For other architectures this test now will not build as there is no implementation of "void test_calculation(void)", but that's ok for me. At least the test is now prepared for extension by others.

ruuddw reviewed (commented):
As the nSIM properties include caches, I suggested EM11D (EM9D has DSP, AGU/XY but no caches). But I believe the Zephyr configuration doesn't include the cache features (EM4_FPDA as basis, not EM6).

ruuddw reviewed (approved):
Thanks for addressing all my feedback!
